File: elixir/bob/bob.exs
defmodule Teenager do
  def hey(input) do
    cond do
      silence?(input) -> "Fine. Be that way!"
      shouting?(input) -> "Whoa, chill out!"
      question?(input) -> "Sure."
      true -> "Whatever."
    end
  end

  defp shouting?(input) do
    Regex.match?(~r/[[:upper:]]/m, input) && Regex.match?(~r/^[[:upper:]0-9\W]+!?$/m, input)
  end

  defp silence?(input) do
    Regex.match?(~r/^\s*$/m, input)
  end

  defp question?(input) do
    String.ends_with?(input, "?")
  end
end

File: ruby/bob/bob.rb
class Bob
  def self.hey(remark)
    case
    when silent?(remark)
      "Fine. Be that way!"

    when shouting?(remark)
      "Whoa, chill out!"

    when asking?(remark)
      "Sure."

    else
      "Whatever."
    end
  end

  def self.silent?(remark)
    remark.strip.empty?
  end

  def self.asking?(remark)
    remark.end_with?("?")
  end

  def self.shouting?(remark)
    remark.upcase == remark && letters?(remark)
  end

  def self.letters?(remark)
    remark.match(/[A-Za-z]+/)
  end
end
